src/libopensc/card-eoi.c

```diff
@@ -0,0 +1,532 @@
+/*
+ * Support for the eOI card
+ *
+ * Copyright (C) 2022 Luka Logar <luka.logar@iname.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#include <stdlib.h>
+#include <string.h>
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#if defined(ENABLE_SM) && defined(ENABLE_OPENPACE)
+
+#include <openssl/aes.h>
+#include <openssl/sha.h>
+#include "opensc.h"
+#include "internal.h"
+#include "sm/sm-eac.h"
+#include "card-eoi.h"
+
+static struct sc_card_operations eoi_ops;
+
+static struct {
+	int len;
+	struct sc_object_id oid;
+} eoi_curves[] = {
+	/* secp384r1 */
+	{384, {{1, 3, 132, 0, 34, -1}}}
+};
+
+/* The description of the driver. */
+static struct sc_card_driver eoi_drv =
+{
+	"eOI (Slovenian eID card)",
+	"eOI",
+	&eoi_ops,
+	NULL, 0, NULL
+};
+
+static const struct sc_atr_table eoi_atrs[] = {
+	/* Contact interface */
+	{ "3b:d5:18:ff:81:91:fe:1f:c3:80:73:c8:21:10:0a", NULL, NULL, SC_CARD_TYPE_EOI, 0, NULL },
+	/* Contactless interface */
+	{ "3b:85:80:01:80:73:c8:21:10:0e", NULL, NULL, SC_CARD_TYPE_EOI_CONTACTLESS, 0, NULL },
+	{ NULL, NULL, NULL, 0, 0, NULL }
+};
+
+/*
+ * CAN is stored encrypted in a file that (looks like) is pointed to by 'Card CAN' PIN object.
+ * eoi_decrypt_can() decrypts CAN from it's encrypted form
+ */
+
+static void rol(u8 *to, const u8 *from)
+{
+	int i;
+	u8 b = from[0] & 0x80;
+	for (i = 15; i >= 0; i--) {
+		u8 bo = b;
+		b = from[i] & 0x80;
+		to[i] = (from[i] << 1) | (bo ? 1 : 0);
+		if ((i == 15) && bo)
+			to[i] = (to[i] ^ 0x87) | 1;
+	}
+}
+
+static int aes256_ecb_encrypt(const u8 *key, const u8 *input, const u8 *output)
+{
+	EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
+	int r = 0, len = AES_BLOCK_SIZE;
+	if (!ctx)
+		goto err;
+	if (!EVP_EncryptInit(ctx, EVP_aes_256_ecb(), key, NULL))
+		goto err;
+	if (!EVP_EncryptUpdate(ctx, (u8 *)output, &len, input, AES_BLOCK_SIZE))
+		goto err;
+	/* Since it's exactly 1 block, no need to call EVP_EncryptFinal */
```

Comment:
nit: I would opt in to call the final even though it will be the no-op just to "follow the OpenSSL API". I think it is never guaranteed that the Update part will return the data if it is full block.
But I do not insist on changing this if others are ok with it.

Reply:
Done. I guess the space and performance overheads are insignificant. Although, at least in my version of OpenSSL, it doesn't like `EVP_EncryptFinal()` when `len = 0`, so I call it only if necessary.